### ISSUE TITLE
Provide an option for Link copy format

### ISSUE THREAD

**Issue by yorkxin** [yorkxin/copy-as-markdown#12]
i.e. Let user choose whether the link is formatted as

```
[Text](url)
```

or

```
[Text][id]

[id]:url
```

**Comment by the-solipsist** [comment]
This feature would be very welcome!

**Comment by malcook** [comment]
if you do this:

another useful format is that of org-mode, which is
`[[url][text]]`

and another is as html link(s)

**Comment by tohagan** [comment]
Ideally I think you'd allow the user to create custom link styles. One solution would be to allow a user to create one or more named link styles (Markdown, Dockuwiki, MediaWiki, HTML ... whatever) and then define a link text template for each style ...

- Markdown = `"[${title}](${link})"`
- Dokuwiki = `" [[${link}|${title}]]"` 
- HTML= `"<a href='{$link}'>${title}</a>"`

The user would then select this named style _when they copy a link_. 

For users like me are likely to want frequently use all the different link styles at different times and so be able to quickly choose the link format when I copy a link (not have to navigate inside the extension options to re-select a link style every time). So if you added this I'd make the link styles names appear on the link menu that appears when you perform the link copy. They could appear be a "checked" menu option that does not close the menu.

So your menu might be ...

- Current Tab
- All Tabs
- Selected Tabs
- ---- Divider -----
- [x] Markdown
- [  ] Dokuwiki
- [  ] HTML 

The options can be hidden if only one style (Markdown) is defined.

**Comment by chitsaou** [comment]
@tohagan Thank you for your suggestion. In fact I sometimes also have need to copy links in non-Markdown format e.g. RTF, in order to paste to rich text editors like Dropbox Paper. I think the UI you proposed is simple and easy to understand. I will consider working on this feature :)

**Comment by tohagan** [comment]
Thanks very much :+1:   I think it will be a really valuable addition and hopefully not that hard to code

**Comment by sebastienbarre** [comment]
This looks great. At this point you are one of the only Copy Link extensions that is multiprocess. Ideally I would love for it to do everything the [CoLT](https://addons.mozilla.org/en-US/firefox/addon/colt/) extension used to do (especially the "Copy Link Text" menu item, which copies a hyperlink's associated text to the clipboard).

**Comment by keisuke-nakata** [comment]
Recent Firefox update destroyed almost all plugins which provide "copy link as something", except for this plugin.
Supporting "copy link as HTML, or any format you specify" feature is very helpful for Firefox users.

**Comment by sgourichon** [comment]
Another overnight-obsoleted extension is [Easy Copy](https://addons.mozilla.org/en-US/firefox/addon/easy-copy/?src=search). 

Easy Copy has been on my shortlist of best extensions for years

It has a simple yet very flexible configuration system which basically allows you to generated whatever markup language you need, while supporting many uses cases. Examples:

* don't select anything, copy -> becomes link to the page, with page title as link text
* _select any snippet of text in the page_, copy -> becomes link to the page, with the text you selected as link text _<<== the one I use most_
* don't select anything, right-click on link, copy -> same result as precisely selecting the link and copying it, just much easier
* the simple and powerful configuration options allow to provide the behavior of [CoLT](https://addons.mozilla.org/en-US/firefox/addon/colt/).
* copy all tabs as links (all this, whatever markup language)

## Screenshots

![use](https://addons.cdn.mozilla.net/user-media/previews/full/124/124227.png?modified=1419676255)

![configuration](https://addons.cdn.mozilla.net/user-media/previews/full/148/148910.png?modified=1419676255)

## Conclusion

* Praise "copy-as-markdown" for being compatible with Firefox 57+!
* I guess it can technically be extended to support all this. If would be soo great!

**Comment by jmfayard** [comment]
Hello @chitsaou 
Just wanted to say that i like your plugin a lot. 
Do you have plans to implement this feature?
That would make it instantly 2 times better.

Please add as well Asciidotor format: http://example.com[Title of the webpage]

**Comment by bhrutledge** [comment]
Thanks for this extension! I frequently use it to copy a bunch of tabs into my Markdown  notes. However, I prefer `-` over `*` for my list bullets. Can that be added as an option?


## Expected Behavior

Set a preference for list bullet to `-`, then "All tabs link" or  "Selected tabs link" would copy:

```
- [chitsaou/copy-as-markdown: Copying Link, Image and Tab(s) as Markdown Much Easier.](https://github.com/chitsaou/copy-as-markdown)
- [Copy as Markdown - Chrome Web Store](https://chrome.google.com/webstore/detail/copy-as-markdown/fkeaekngjflipcockcnpobkpbbfbhmdn?hl=en)
- [Copy as Markdown – Get this Extension for 🦊 Firefox (en-US)](https://addons.mozilla.org/en-US/firefox/addon/copy-as-markdown/)
```

## Actual Behavior

```
* [chitsaou/copy-as-markdown: Copying Link, Image and Tab(s) as Markdown Much Easier.](https://github.com/chitsaou/copy-as-markdown)
* [Copy as Markdown - Chrome Web Store](https://chrome.google.com/webstore/detail/copy-as-markdown/fkeaekngjflipcockcnpobkpbbfbhmdn?hl=en)
* [Copy as Markdown – Get this Extension for 🦊 Firefox (en-US)](https://addons.mozilla.org/en-US/firefox/addon/copy-as-markdown/)
```

**Comment by roryokane** [comment]
Like @bhrutledge, I would prefer my copied text use `-` as the list bullet instead of `*`. I don’t have time right now to try to edit this extension, but here are some notes towards developing that feature.

This is the function that writes `*` as the list bullet:

https://github.com/yorkxin/copy-as-markdown/blob/e990a320697486108504d88ddee7d676eac8abb9/src/background/markdown.js#L38-L40

The simplest edit to that file that would support customizing the list bullet and nothing else might be this:

```javascript
export function list(theList) {
  const bulletCharacter = userOptions.bulletCharacter || '*';
  return theList.map((item) => `${bulletCharacter} ${item}`).join('\n');
}
```

Then [src/lib/options-manager.js](https://github.com/yorkxin/copy-as-markdown/blob/e990a320697486108504d88ddee7d676eac8abb9/src/lib/options-manager.js) would need to be customized to support a `bulletCharacter` option, and [src/ui/options.js](https://github.com/yorkxin/copy-as-markdown/blob/e990a320697486108504d88ddee7d676eac8abb9/src/ui/options.js) and [src/ui/options.html](https://github.com/yorkxin/copy-as-markdown/blob/e990a320697486108504d88ddee7d676eac8abb9/src/ui/options.html) would have to be edited to support the user editing that option. The editing interface would be either radio buttons or a pop-up menu. The buttons or menu would have two options: `-` and `*`.

The template-based customization feature described in [tohagan’s comment](https://github.com/yorkxin/copy-as-markdown/issues/12#issuecomment-228279016) sounds more useful than just being able to customize the bullet character, and it would solve my problem. However, it sounds more complicated to implement.

**Comment by roryokane** [comment]
As an alternative to adding an bullet customization option as I described above, we could change the hard-coded bullet character from `*` to `-` by editing the `list` function. Do any current users of the extension (including its author) actively prefer `*` over `-`, or was `*` an arbitrary choice? I prefer `-` for various reasons, including that it matches [Prettier](https://prettier.io/)’s Markdown formatting.

**Comment by yorkxin** [comment]
Related: #98

**Comment by edrex** [comment]
Thanks for maintaining this extension for so long yorkxin :heart: 

I go back and forth between org-mode and markdown, and as others have noted (@tohagan @malcook back in 2016) there are a few other text-based formats in use. A custom format option with a template field would be great. Concat multiple with newlines. Otherwise we could just add the requested formats, but it would make the options more overwhelming for the large majority of users. Since I need this I'll have a go and PR if that's cool.

**Comment by yorkxin** [comment]
Hi @edrex (and everyone), I understand this feature has been long asked for, ever since the birthday of this extension.

I have a prototype of custom formats, but I am trying to make it as simple as possible, so that people who are not tech-savvy don't get scared. This requires some UX design efforts. I am working with a UX designer to figure it out. 

Meanwhile, I'll explore some options and see if I can ship a simplest version.

Thanks for your patient!

**Comment by yorkxin** [comment]
Implemented #158 . Released v3.1.0. Submitted to Chrome & Edge store for review. It is now available on Firefox add-ons. Please try it out when it is released.

**Comment by yorkxin** [comment]
Version 3.1.0 is now available on Chrome web store. If you get the extension from there, please update and try it out!

I'll close this issue for now.

Thanks for your patience for 12 years!
